### PR TITLE
sql function regex_match (and some tests for other sql functions)

### PR DESCRIPTION
**Pull request by directionless**
This creates a `regexp_match` sql function, tests, and associated docs.

Add tests for more of the string sql functions

### CODE REVIEW

File: osquery/sql/sqlite_string.cpp [hunk 1]
```diff
@@ -122,6 +122,51 @@ static void regexStringSplitFunc(sqlite3_context* context,
...
+  boost::smatch results;
+ 
+  bool isMatchFound = boost::regex_search(input, results, pattern);
+
+  if(!isMatchFound) {
+    sqlite3_result_null(context);
+    return;
+  }
+
+  if( index > results.size() ) {
```
Comment (fmanco) [inline osquery/sql/sqlite_string.cpp]:
```suggestion
  if( index >= results.size() ) {
```

File: osquery/sql/sqlite_string.cpp [hunk 2]
```diff
@@ -122,6 +122,51 @@ static void regexStringSplitFunc(sqlite3_context* context,
...
+    sqlite3_result_null(context);
+    return;
+  }
+
+  auto result  = results[index].str();
```
Comment (fmanco) [inline osquery/sql/sqlite_string.cpp]: Don't create a copy here just access this directly on the line bellow.

File: osquery/sql/sqlite_string.cpp [hunk 3]
```diff
@@ -122,6 +122,51 @@ static void regexStringSplitFunc(sqlite3_context* context,
...
+  
+  // const char * result =  results[index].str().c_str();
+
+  sqlite3_result_text(context,
+		      result.c_str(),
```
Comment (fmanco) [inline osquery/sql/sqlite_string.cpp]:
```suggestion
		      results[index].str().c_str(),
```

**Comment by fmanco** [comment]
@directionless thanks for the PR, this is a great feature. I left a few comments. Sorry for the delay reviewing this.

File: osquery/sql/sqlite_string.cpp [hunk 4]
```diff
@@ -122,6 +122,51 @@ static void regexStringSplitFunc(sqlite3_context* context,
...
+      SQLITE_NULL == sqlite3_value_type(argv[1]) ||
+      SQLITE_NULL == sqlite3_value_type(argv[2])) {
+    sqlite3_result_null(context);
+    return;
+  }
+
+  // parse and verify input parameters
+  std::string input((char*)sqlite3_value_text(argv[0]));
+  boost::regex pattern((char*)sqlite3_value_text(argv[1]), boost::regex::extended);
+  auto index = static_cast<size_t>(sqlite3_value_int(argv[2]));
```
[the remark this answers is not in the document]
Comment (directionless) [reply osquery/sql/sqlite_string.cpp]: Hrm. It's unsigned, -1 is `18446744073709551615`. And strings map to `0`. I'm not sure what the right limits to place here would be

File: osquery/sql/sqlite_string.cpp [hunk 5]
```diff
@@ -122,6 +122,51 @@ static void regexStringSplitFunc(sqlite3_context* context,
...
+  if (SQLITE_NULL == sqlite3_value_type(argv[0]) ||
+      SQLITE_NULL == sqlite3_value_type(argv[1]) ||
+      SQLITE_NULL == sqlite3_value_type(argv[2])) {
+    sqlite3_result_null(context);
+    return;
+  }
+
+  // parse and verify input parameters
+  std::string input((char*)sqlite3_value_text(argv[0]));
+  boost::regex pattern((char*)sqlite3_value_text(argv[1]), boost::regex::extended);
```
[the remark this answers is not in the document]
Comment (directionless) [reply osquery/sql/sqlite_string.cpp]: 👍 I added some catches to my regex_match, as well as to the existing regex_split

**Comment by directionless** [comment]
@fmanco Thanks so much for the review! I think I got most of the requested changes. I'm not sure what we should do with the index parameter. I can add some checks to all three functions, but something is translating it to an unsigned int. So we'd have to decide on reasonable bounds

**Comment by packetzero** [comment]
You already check index <= size, that should be sufficient

**Comment by JPerkster** [comment]
@directionless Would you by chance have any further updates here? This is a great feature that would be extremely useful

**Pull request merged**